cmd/argoexec/commands/emissary.go

```diff
@@ -0,0 +1,216 @@
+package commands
+
+import (
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	"github.com/argoproj/argo-workflows/v3/util/archive"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+	"github.com/argoproj/argo-workflows/v3/workflow/util/path"
+)
+
+var (
+	varArgo       = "/var/run/argo"
+	containerName = os.Getenv("ARGO_CONTAINER_NAME")
+	template      = &wfv1.Template{}
+)
+
+func NewEmissaryCommand() *cobra.Command {
+	return &cobra.Command{
+		Use:          "emissary",
+		SilenceUsage: true, // this prevents confusing usage message being printed when we SIGTERM
+		RunE: func(cmd *cobra.Command, args []string) error {
+			exitCode := 64
+
+			defer func() {
+				err := ioutil.WriteFile(varArgo+"/ctr/"+containerName+"/exitcode", []byte(strconv.Itoa(exitCode)), 0600)
+				if err != nil {
+					println(fmt.Errorf("failed to write exit code: %w", err))
+				}
+			}()
+
+			// this also indicates we've started
+			if err := os.MkdirAll(varArgo+"/ctr/"+containerName, 0700); err != nil {
+				return fmt.Errorf("failed to create ctr directory: %w", err)
+			}
+
+			name, args := args[0], args[1:]
+
+			signals := make(chan os.Signal, 1)
+			defer close(signals)
+			signal.Notify(signals)
+			defer signal.Reset()
+			go func() {
+				for s := range signals {
+					if s != syscall.SIGCHLD {
+						_ = syscall.Kill(-os.Getpid(), s.(syscall.Signal))
+					}
+				}
+			}()
+
+			data, err := ioutil.ReadFile(varArgo + "/template")
+			if err != nil {
+				return fmt.Errorf("failed to read template: %w", err)
+			}
+
+			if err := json.Unmarshal(data, template); err != nil {
+				return fmt.Errorf("failed to unmarshal template: %w", err)
+			}
+
+			name, err = path.Search(name)
+			if err != nil {
+				return fmt.Errorf("failed to find name in PATH: %w", err)
+			}
+
+			command := exec.Command(name, args...)
+			command.Env = os.Environ()
+			command.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+			stdout, err := os.Create(varArgo + "/ctr/" + containerName + "/stdout")
```

Comment:
I'm aware we only need to capture stdout for (a) main containers and (b) if there our outputs - v1.1

Reply:
could use `ioutil.Discard` for that